docs/js-guide.md

```diff
@@ -0,0 +1,143 @@
+   feature without excessive overhead.
+
+   _Some language features (generators, as a historical example) are difficult
+   to transpile for older browsers without introducing a lot of code that
+   slows startup and makes debugging difficult._
+
+As an exception, [JSX](https://reactjs.org/docs/introducing-jsx.html) is used for
+creating UI components.
+
+### Variable declarations
+
+Variables should be declared with `const` where possible, as this makes it clear
+that the variable is only assigned in one place.
+
+## Formatting code
+
+All of our projects use [Prettier](https://prettier.io) to format JavaScript code,
+with the following `.prettierrc` configuration:
+
+```
+{
+  "arrowParens": "avoid",
+  "singleQuote": true
+}
+```
+
+All other config options are set to the Prettier defaults.
+
+## Lint setup
+
+Our projects use [ESLint](https://eslint.org) to:
+
+- Catch common JavaScript coding errors
+- Detect HTML usage errors in Preact components
+- Identify potential accessibility errors in Preact components
+- Ensure correct usage of the "hooks" APIs in Preact components
+- Encourage stylistic consistency across projects (although many basic aspects
+  of code style are handled by Prettier)
+
+Projects should use the `eslint-config-hypothesis` package from this repository
+as a base configuration, with any additional project-specific configuration added
+on top.
+
+[Lint suppressions](https://eslint.org/docs/2.13.1/user-guide/configuring#disabling-rules-with-inline-comments)
+may be used where there are reasonable reasons to ignore a particular lint warning.
+Try to scope any suppressions to the narrowest scope possible and add a comment
+to explain any that may be non-obvious to another developer reading the code
+in future.
+
+## Documenting and type-checking code
+
+### JSDoc
+
+It is recommended to use [JSDoc](https://jsdoc.app) to document function
+parameters and object properties.
+
+Hypothesis uses the variant of JSDoc understood by the TypeScript compiler.
+See [the reference](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html)
+in the TypeScript handbook for guidance on supported JSDoc syntax as well as
+details on how to specify various types correctly.
+
+### Typechecking
+
+We are investigating the use of the TypeScript compiler for some
+projects to check the semantic correctness of code based on JSDoc annotations
+and provide a better development experience in IDEs.
+In these projects a [tsconfig.json file](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)
+with the `allowJs` and `checkJs` settings is used.
+
+When typechecking is used for a project, expressions that are difficult to
+check can be ignored using `// @ts-ignore` comments. As with ESLint suppressions,
+try to scope these to the smallest possible scope.
+
+### prop-types
+
+The [prop-types](https://reactjs.org/docs/typechecking-with-proptypes.html) package
```

Comment:
This section jumps into the assumption that we're talking about preact components. Perhaps be a little more explicit about that?

Reply:
Ah, yes. Good point.